### ISSUE TITLE
UIFrameFlash taints Communities and Spellbook UIs

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#183]
The UIFrameFlash animation system has been a spreader of taint since patch 5.2 (~2013), leading to numerous "interface action blocked" messages within the default UI **without addons even having to directly calling any UIFrameFlash-related function**.

An example of one such issue this causes in the Retail client is that in rare circumstances opening and closing the Spellbook can trigger a blocked action message as it attempts to show or hide action buttons based upon the "Always Show Actionbars" setting.

A common way of accidentally spreading taint through the UIFrameFlash system is by installing a chat message event filter, which [will taint execution](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/ChatFrame.lua#L3534-L3544) upon processing an event, leading into a call to [FCF_StartAlertFlash](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/ChatFrame.lua#L3905) which itself calls [UIFrameFlash](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/FloatingChatFrame.lua#L1486-L1494).

This issue - to varying extents - affects all current live and test clients of each client flavour:

- Retail (9.1.0 and 9.1.5)
- BCC (2.5.2)
- Classic Era (1.13.7 and 1.14.0)

A reproducible example of the issue is provided below. While this particular example is somewhat contrived, it is important to note that [UIFrameFlash is used by many parts of the default UI](https://www.townlong-yak.com/globe/wut/#q:UIFrameFlash), and one problematically one usage is deep within [the communities UI](https://github.com/Gethe/wow-ui-source/blob/e337b8949ffad2876ea0489d8331db2414342d32/AddOns/Blizzard_Communities/CommunitiesList.lua#L541-L546) which leads to a significant amount of taint permeating that system.

---

The easiest way to reproduce the issue on the Retail client is as follows:

1. Log into a max-level character.
2. Go to the Interface options panel and untick the "Always Show Actionbars" setting in "ActionBars".
3. Click "Okay" and close the game menu.
4. Run the following command:
```
/run UIFrameFlash(CreateFrame("Frame"), 0.1, 0.1, -1)
```
5. Open the Talent UI and change your active specialization.
6. **Enter combat with an enemy.**
7. Open the Spellbook UI.

At this point you'll receive an "Interface action failed because of an AddOn" message, and any actionbuttons on your bar which were hidden due to not having spells placed will not have been shown. A full stack trace of what occurs can be found below.

```
1x [ADDON_ACTION_BLOCKED] AddOn '*** ForceTaint_Strong ***' tried to call the protected function 'MultiBarBottomRightButton1:Hide()'.
[string "=[C]"]: in function `Hide'
[string "@FrameXML\ActionButton.lua"]:481: in function `HideGrid'
[string "@FrameXML\MultiActionBars.lua"]:139: in function `MultiActionBar_UpdateGrid'
[string "@FrameXML\MultiActionBars.lua"]:121: in function `MultiActionBar_ShowAllGrids'
[string "@FrameXML\SpellBookFrame.lua"]:191: in function <FrameXML\SpellBookFrame.lua:181>
[string "=[C]"]: in function `Show'
[string "@FrameXML\UIParent.lua"]:2956: in function `SetUIPanel'
[string "@FrameXML\UIParent.lua"]:2801: in function `ShowUIPanel'
[string "@FrameXML\UIParent.lua"]:2669: in function <FrameXML\UIParent.lua:2665>
[string "=[C]"]: in function `SetAttribute'
[string "@FrameXML\UIParent.lua"]:3520: in function `ShowUIPanel'
[string "@FrameXML\SpellBookFrame.lua"]:75: in function `ToggleSpellBook'
[string "TOGGLESPELLBOOK"]:1: in function <[string "TOGGLESPELLBOOK"]:1>
```

The chain of events that causes this action blocked message is the following:

1. We insecurely call UIFrameFlash and trigger an insertion [into the global FLASHFRAMES table](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L4024).
2. Upon changing specializations - or learning any new spells - the `LEARNED_SPELL_IN_TAB` event fires a few times, [causing `SpellBookFrame.flashTabs` to be set](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/SpellBookFrame.lua#L146-L155).
3. Upon showing the Spellbook we [enter its OnShow handler](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/SpellBookFrame.lua#L181-L191) which, due to `flashTabs` being set, will lead to a call into UIFrameFlash.
4. UIFrameFlash [iterates over the contents of `FLASHFRAMES`](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L3987-L3996) to check if it's about to insert a duplicate reference to a frame. **As we insecurely inserted an entry here earlier, this taints execution.** This is similar to issue #169.
5. Upon returning from UIFrameFlash execution is tainted and we enter MultiActionBar_ShowAllGrids, which leads to a call to either ActionButton:ShowGrid or ActionButton:HideGrid, which [attempts to show or hide action buttons](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/ActionButton.lua#L469-L481) while tainted. As we're in combat at this time, this leads to an action blocked error.

---

A simple fix with a healthy dollop of `securecall` would be as follows:

1. [Take a page out of Classic's book](https://github.com/Gethe/wow-ui-source/blob/6c7cc51788340fd5c07770429bc00f8108955800/Interface_TBC/FrameXML/UIParent.lua#L2940-L2943) and [securecall the test to see if a given frame exists in FLASHFRAMES](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L3990-L3996).
2. The UIFrameIsFlashing function is called by a few areas of the chatframe code and additionally needs fixing in a similar manner to the above.
3. The [call to `tDeleteItem` in UIFrameFlashStop](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L4089-L4090) will also iterate over the contents of `FLASHFRAMES` and taint execution, so this additionally needs to be wrapped in a securecall.
4. In UIFrameFlash_OnUpdate, [this initial loop over UIFrameFlashTimers](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L4035-L4038) should be `secureexecuterange`'d. This is because taint from individual entries in that table can taint the assignment of updated time values for other initially-secure entries, and this can trigger [a tainted read in UIFrameFlash](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L4000-L4003) if a `syncId` is specified - and those are heavily used by the FloatingChatFrame code.
5. (Bonus) Port the fix [for UIFrameFade](https://github.com/Gethe/wow-ui-source/blob/6c7cc51788340fd5c07770429bc00f8108955800/Interface_TBC/FrameXML/UIParent.lua#L2940-L2943) to Retail as well, since that hasn't happened yet. 
    5a. Also apply similar fixes mentioned in steps 2 and 3 to [UIFrameIsFading](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L3972-L3979) and [UIFrameRemoveFade](https://github.com/Gethe/wow-ui-source/blob/fe3a9d7991a529fb826a400d562e3964ef785a15/FrameXML/UIParent.lua#L3912-L3914), and port those fixes to Classic. These functions are also called by the FloatingChatFrame code in a few locations.

As UIFrameFlash is just an animation system there should be no security impact from just securecall-ing these few bits.

**Comment by Meorawr** [comment]
This is going to be looked at in a future patch. The UIFrameFlash system itself _may_ relegated to deprecation and all usages of it internally removed, so any addons that happen to be using it currently should probably consider migrating to the widget animation system instead.

**Comment by Meorawr** [comment]
Speculatively marking this as fixed - 10.2.7 has made changes to UIFrameFlash/Fade which incorporate a bunch of securecall barriers at previously problematic points.

Haven't tested thoroughly, but can raise new tickets if it continues to be a pain.
